### PR TITLE
First draft of a short article pushing away from using `can_connect()`

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
When happy with this, it would presumably go to discourse.charmhub.io.

Also when happy with the general approach, the same types of code can be used in the tutorials, etc.

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@benhoyt do you have access to review here? I can add you if not. This is still a draft, but thoughts on it so far would be great if you have some time.

**Comment by benhoyt** [comment]
@tonyandrewmeyer Looks like I do have access to review here -- thanks. I'll try to take a look today.

File: charm_temp_errors.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# Handling Temporary Errors in Charms
+
+There are a few types of errors that can occur while a Charm is handling an event that are expected to resolve within a fairly short timeframe without human intervention, and can occur at any point during the event handling, despite Juju's efforts to provide a stable snapshot of the model state.
```
Comment (benhoyt) [inline charm_temp_errors.md]: This is a bit hard to parse. It sounds like the *event* is expected to resolve, but I think that's referring to the *errors*. Can you restructure/simplify/split up this sentence?

Also, nit: Charm -> charm

```suggestion
There are a few types of errors that can occur while a charm is handling an event that are expected to resolve within a fairly short timeframe without human intervention, and can occur at any point during the event handling, despite Juju's efforts to provide a stable snapshot of the model state.
```

File: charm_temp_errors.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+If Pebble can't be reached, then this will raise a `ConnectionError` and the charm will enter an error state. In some cases, that might be the right thing to do! However, we'd rather avoid the error state unless human intervention is required to fix the problem, and we're confident that in most cases this is actually just a temporary issue and trying again after a short time will solve the problem.
+
+A common pattern is to check Pebble first, and defer if it's not available:
+
+```python
+if container.can_connect():
+    container.replan()
+else:
+    event.defer()
```
Comment (benhoyt) [inline charm_temp_errors.md:32]: I'm afraid that people might read up to here and copy and paste this pattern. :-) Can we have a red "not like this" warning or at least a `# DON'T DO THIS -- SEE BELOW` type of comment?

Update: I see after reading further you're telling a story, and that's good. But I still think it'll be too easy for people to not read (people, myself included, do a lot of not-reading) and think this, or the next code snippet, is what you're suggesting... thoughts?

Also, probably better to restructure as `if not container.can_connect()` with an early return, keeping the `container.replan()` at the happy-path nesting level, as I think we/charmers would normally write it.

File: charm_temp_errors.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    return
+```
+
+A problem with this approach is that `can_connect()` doesn't guarantee that Pebble will be reachable the next time the charm tries to communicate with it - just that it was at that point in time. This means that we've introduced a race condition where the `can_connect()` can succeed but the `replan()` fails, and we again end up in an error state. Let's fix that:
```
Comment (benhoyt) [inline charm_temp_errors.md]:
```suggestion
A problem with this approach is that `can_connect()` doesn't guarantee that Pebble will be reachable the next time the charm tries to communicate with it - just that it was at that point in time. This means that we've introduced a race condition where the `can_connect()` can succeed but the `replan()` fails, and we again end up in an error state. Let's try to fix that:
```

File: charm_temp_errors.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        container.replan()
+    except ops.pebble.ConnectionError:
+        event.defer()
+        return
+else:
+    event.defer()
+    return
+```
+
+However, it's clear now that the `can_connect()` call isn't doing anything useful - it's pinging Pebble but most of the time we expect Pebble will be available (so it's not required) and when it's not we already have the code to handle that. Let's simplify the code:
```
Comment (benhoyt) [inline charm_temp_errors.md]:
```suggestion
However, it's clear now that the `can_connect()` call isn't doing anything useful - it's pinging Pebble but most of the time we expect Pebble will be available (so it's not required) and when it's not we already have the `try`/`except` to handle that. Let's make an attempt to simplify the code:
```

File: charm_temp_errors.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+We're pretty sure that when we do get this failure, it's a temporary problem - perhaps there's high load, or some temporary networking issues. One issue with using `defer()` is that we don't know, or have any control over, when the event handler will be retried. `defer()` also restarts the entire handler, so any work we've already done will get repeated, which is wasteful. Let's adjust the code so that we give Pebble a few chances:
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY ** attempt)
+    else:
```
Comment (benhoyt) [inline charm_temp_errors.md]: I think it's clearer/simpler to remove the `else: break` (especially with the `else` below) and just put a `break` after the `container.replan()` line.

File: charm_temp_errors.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+We're pretty sure that when we do get this failure, it's a temporary problem - perhaps there's high load, or some temporary networking issues. One issue with using `defer()` is that we don't know, or have any control over, when the event handler will be retried. `defer()` also restarts the entire handler, so any work we've already done will get repeated, which is wasteful. Let's adjust the code so that we give Pebble a few chances:
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY ** attempt)
```
Comment (benhoyt) [inline charm_temp_errors.md]: This is a nice trick, but a little odd if `PEBBLE_RETRY_DELAY` is less than 1, say 0.5 (seconds). Then it'll go 1.0, 0.5, 0.25, etc.

File: charm_temp_errors.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    else:
+        break 
+else:
+    logger.error("Unable to reach Pebble after %.1f seconds.", too_lazy_to_do_the_maths)
+    raise SystemExit(1)
```
Comment (benhoyt) [inline charm_temp_errors.md]: Would this be better as `raise Exception("Unable to reach Pebble after X seconds")`?

File: charm_temp_errors.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+The only time we aren't expecting Pebble to be available is before we get the container's `PebbleReadyEvent`. If our code is running in an event handler that's potentially before then (see [a charm's lifecycle](https://juju.is/docs/sdk/charm-lifecycle)), we'd want to have the work be handled by a later event (or defer). If the code is running in a handler than could occur in the "setup" as well as "operation" or "teardown" phases, then we might prefer to always avoid the error state.
+
+Let's put it all together into a simple, reusable, method:
+
+.. TODO: I don't love how this is. A decorator or context manager is likely best. Maybe the handling of defer, return etc is too much? A class with a __getattr__ could make the call look much more like the normal version.
```
Comment (benhoyt) [inline charm_temp_errors.md]: Yeah, I'm not sure I love the `tolerant_pebble` proposal either. It seems to do too much in one tool, really two separate things: the retrying of the Pebble call, and the handling of an error. Might be worth a quick voice discussion to brainstorm whether we want this or how it should look.

**Review by benhoyt — CHANGES_REQUESTED** [review]
I originally started reviewing this as if it was going somewhere under https://juju.is/docs/sdk, but then I re-read the PR description and realised it was intended for Discourse, where the less formal style makes a lot more sense. So take some of my comments in that vein.

File: charm_temp_errors.md (hunk 2 above)
Comment (tonyandrewmeyer) [reply charm_temp_errors.md:32]: > I'm afraid that people might read up to here and copy and paste this pattern. :-) Can we have a red "not like this" warning or at least a `# DON'T DO THIS -- SEE BELOW` type of comment?
> 
> Update: I see after reading further you're telling a story, and that's good. But I still think it'll be too easy for people to not read (people, myself included, do a lot of not-reading) and think this, or the next code snippet, is what you're suggesting... thoughts?

Agreed. I've put one in above, but I suspect it's not going to properly render in Discourse (I don't think Commonmark or Discourse have any built-in support for admonition blocks - you can do bbcode blocks but then you have to have custom styles so the theme really needs to support it). I'll adjust to whatever looks reasonable when actually posting.

> Also, probably better to restructure as `if not container.can_connect()` with an early return, keeping the `container.replan()` at the happy-path nesting level, as I think we/charmers would normally write it.

Yes, good call - done.

Comment (benhoyt) [reply charm_temp_errors.md:32]: This looks good. However, if it doesn't render on Discourse, maybe we can just keep it simple with bold or something: **Warning: don't use this code ...**

File: charm_temp_errors.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+        break
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY * attempt)
+else:
+    raise RuntimeError("Unable to reach Pebble")
```
Comment (benhoyt) [inline charm_temp_errors.md]: With this approach we lose the original exception. Could we instead re-raise the exception, or at least include the original exception, in the final one?

```python
for attempt in range(MAX_PEBBLE_RETRIES):
    try:
        container.replan()
        break
    except ops.pebble.ConnectionError:
        if attempt == MAX_PEBBLE_RETRIES - 1:
            raise
        time.sleep(PEBBLE_RETRY_DELAY * attempt)
```

**Review by benhoyt — APPROVED** [review]
Couple of minor suggestions, but I'm pretty happy with this for a discourse post -- thanks!

File: charm_temp_errors.md (hunk 9 above)
Comment (tonyandrewmeyer) [inline charm_temp_errors.md]:
```suggestion
for attempt in range(MAX_PEBBLE_RETRIES):
    try:
        container.replan()
        break
    except ops.pebble.ConnectionError:
        if attempt >= MAX_PEBBLE_RETRIES - 1:
            raise
        time.sleep(PEBBLE_RETRY_DELAY * attempt)
```
